app/Http/Requests/StoreImageAnnotation.php

```diff
@@ -50,7 +50,59 @@ public function withValidator($validator)
         $validator->after(function ($validator) {
             if ($this->input('shape_id') === Shape::wholeFrameId()) {
                 $validator->errors()->add('shape_id', 'Image annotations cannot have shape WholeFrame.');
+                return;
             }
+
+            if (!$this->input('points')) {
+                $validator->errors()->add('points', 'Invalid shape. No points given.');
+                return;
+            }
+
+            $fails = false;
+            $points = $this->input('points');
+
+            switch ($this->input('shape_id')) {
+                case (Shape::circleId()):
+                    // Circle needs x,y coordinate and non zero radius
+                    $fails = count($points) !== 3 || intval(($points)[2]) === 0;
+                    break;
+                case (Shape::ellipseId()):
+                case (Shape::rectangleId()):
+                    // Ellipse and rectangle need exactly 4 distinct points
+                    $fails = count($points) !== 8 || $this->countDistinctPoints($points) !== 4;
+                    break;
+                case (Shape::lineId()):
+                    // Line need at least 2 distinct points
+                    $fails = count($points) < 4 || $this->countDistinctPoints($points) < 2;
+                    break;
+                case (Shape::polygonId()):
+                    // Polygon needs at least 3 distinct points
+                    $fails = count($points) < 6 || $this->countDistinctPoints($points) < 3;
+                    break;
+                case (Shape::pointId()):
+                    // Points needs exactly one x and y coordinate
+                    $fails = count($points) !== 2;
+            }
+
+            if ($fails) {
+                $validator->errors()->add('points', 'Invalid shape.');
+            }
+
```

Comment:
The points validation is already done in the [HasPointsAttribute](https://github.com/biigle/core/blob/95903302705a34a4b2156c93351669352df3f5db/app/Traits/HasPointsAttribute.php#L16-L52) trait. You can probably update the validation there with the additional stuff you added here and it should work everywhere (your tests will tell you). This will also work with the "update" API endpoints that you didn't consider here.

Reply:
It just came into my mind that the trait with the validation method was probably created before I used form requests with validation. This is why it is not located where you would expect validation logic. If you want to clean up the code, you could also implement the point validation as a custom validation rule that can be reused in the store/update requests.